src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java

```diff
@@ -164,7 +164,8 @@ public static void loadNative() {
         }
 
         // Calls disconnect() to cleanup the native part of the wrapper.
-        Cleaner.create().register(this, this::disconnect);
+        Cleaner.create().register(this,
+            () -> PKCS11.disconnect(pNativeData));
```

Comment:
If I'm not mistaken each new call to Cleaner.create() will create a new cleaner and a new deamon thread for it, so it's not recommended to create one cleaner per object.
Also it might be worth double checking that the lambda created here doesn't capture `this`: IIRC there were some subtle cases where a lambda could unexpectedly capture `this`.

Also probably the cleaner itself can't be GC'ed while its thread is running but you might be relying on undocumented behavior. It would be more prudent to stick it in a static variable to retain a strong reference.

Reply:
Good points.  I may change to use a static method instead in the next commit.

Reply:
The new Cleaner code for PasswordCallbackHandler in P11KeyStore.java also creates a new Cleaner instance per object.

Possible capture of 'this' is one reason I find `Runnable`s written out as static classes easier reason about than lambdas.